Fix: Refactor debugging protocol

molant opened the pull request:
<!--

Read our pull request guide:
https://webhint.io/docs/contributor-guide/getting-started/pull-requests/

For the following items put an "x" between the square brackets
(i.e. [x]) if you completed the associated item.

-->

## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- ~~Added/Updated related documentation.~~
- [x] Added/Updated related tests.

## Short description of the change(s)

Refactor of the debugging protocol code. We were having a few issues and needed some clean up.

Fix #1398 

Keep in mind this is a work in progress. I still need to make all the test pass and I want to add some documentation on how the debugging protocol works because I will completely forget in 2 weeks.

<!--

If this is a non-trivial change, include information such as what
benefits this change brings as well as possible drawbacks.

If this fixes an existing issue, include the relevant issue number(s).

Thank you for taking the time to open this PR!

-->

molant commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -22,26 +22,257 @@ import { Crdp } from 'chrome-remote-debug-protocol';
+    loadingFailed = 'loadingFailed'
+}
+
+type NetworkEvent = Crdp.Network.RequestWillBeSentEvent |
```
@antross can you take a 👀 at this? Not sure this is the best option.

antross replied:
Taking a look now...

sarvaje commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -22,26 +22,257 @@ import { Crdp } from 'chrome-remote-debug-protocol';
+    Crdp.Network.LoadingFinishedEvent |
+    Crdp.Network.LoadingFailedEvent;
+
+let overrideInvalidCert: false;
```
IMO, this should be a private variable in each class, and it should be pass through the constructor.

sarvaje commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -22,26 +22,257 @@ import { Crdp } from 'chrome-remote-debug-protocol';
+
+let overrideInvalidCert: false;
+
+class RequestResponse {
```
Each class should go to a different file.

sarvaje commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -22,26 +22,257 @@ import { Crdp } from 'chrome-remote-debug-protocol';
+    }
+
+    private _rawResponse: Buffer | undefined;
+    private getRawResponse() {
```
don't you want this to be `private get rawResponse()` as the other properties??

molant replied:
I wish I could, but it has some `async` stuff and the return type is `NetworkData` instead of `Buffer`

sarvaje replied:
Ahhh, ok! Maybe can you add the type to the function then. `private getRawResponse(): NetworkData`

sarvaje commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -22,26 +22,257 @@ import { Crdp } from 'chrome-remote-debug-protocol';
+            return final;
+        }, {} as Crdp.Network.Headers);
+
+        return this.fetchContent(responseUrl, validHeaders)
```
any reason to not use `async/await`?

molant replied:
so the `catch` can be handle up by whoever is calling this method. This way I don't have to put a `try/catch` here.

sarvaje replied:
ok!

sarvaje commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -22,26 +22,257 @@ import { Crdp } from 'chrome-remote-debug-protocol';
+    }
+
+    private _response: Response | undefined;
+    public getResponse(element: AsyncHTMLElement | null): Response | undefined {
```
`public get response()`?

sarvaje commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -295,14 +489,25 @@ export class Connector implements IConnector {
 
         // DOM is not ready so we queue up the event for later
         if (!this._dom) {
+            console.error(`(${params.requestId}) Queuing onLoadingFailed`);
```
use `logger.error` or remove `console.error`

molant replied:
Remove `console.error`. Left over from testing.

sarvaje commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -503,7 +678,8 @@ export class Connector implements IConnector {
             // DOM is not ready so we queue up the event for later
             if (!this._dom) {
-                this._pendingResponseReceived.push(this.onResponseReceived.bind(this, params));
+                console.error(`(${params.requestId}) Queuing onLoadingFinished`);
```
use `logger.error` or remove `console.error`

molant commented:
@sarvaje @antross I should have addressed all the feedback you gave me in different channels and tests for `connector-chrome` are now passing in my machine so this should be ready for another review.

molant commented:
> now passing in my machine so this should be ready for another review

And for whatever reason the favicon tests fail in Travis 😭

molant commented:
> And for whatever reason the favicon tests fail in Travis 😭

@alrra could it be that because we are using `npm` I'm getting the previous version of `utils-debugging-protocol` and that could be messing things up?

alrra commented:
> we are using npm

We are using [`yarn`](https://travis-ci.org/webhintio/hint/jobs/442486032#L648). 

`npm` is only used when releasing packages.

molant commented:
The problem seems to be that chrome will not download the favicon when running in headless mode.

I think that I'll disable those tests via `is-ci` if needed. Any objections?

antross commented on packages/utils-debugging-protocol-common/src/RequestResponse.ts:
```diff
@@ -0,0 +1,271 @@
+    private _willBeSent: Crdp.Network.RequestWillBeSentEvent | undefined;
+    /** The payload returned by the debugging protocol in the `RequestWillBeSent` event. */
+    public get willBeSent(): Crdp.Network.RequestWillBeSentEvent {
+        return this._willBeSent!;
```
Is there a reason this one is assumed not to be `undefined` whereas the others may be?

molant replied:
Nope, left over from when I was doing tests.

antross commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -136,7 +127,7 @@ export class Connector implements IConnector {
              */
             const query: string = `[src$="${basename}" i],[href$="${basename}" i],[src$="${decodeURIComponent(basename)}" i],[href$="${decodeURIComponent(basename)}" i]`;
-            const newElements: Array<AsyncHTMLElement> = await this._dom.querySelectorAll(query);
+            const newElements: Array<AsyncHTMLElement> = await this._dom!.querySelectorAll(query);
```
Nit: `this._dom!` is technically safe as all callers (or callers of callers) ensured it existed before we got here. However, we're so far removed by this point it's hard to tell without reading a fair amount of code. 

What do you think about adding `dom: CDPAsyncHTMLDocument` as a parameter to this function and having it explicitly passed down from the callers that performed the check to ensure it existed?

antross commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -311,214 +258,30 @@ export class Connector implements IConnector {
 
         const request: Request = {
-            headers: params.response.requestHeaders as HttpHeaders,
+            headers: normalizeHeaders(requestResponse.responseReceived!.response.requestHeaders) as HttpHeaders,
```
Nit: similar to my earlier comment about `this._dom`, consider passing `responseReceived` down as a parameter from the calling functions that checked to ensure it existed to make the fact that this is not `undefined` clearer (and understandable by the compiler).

sarvaje commented:
>The problem seems to be that chrome will not download the favicon when running in headless mode.
>
>I think that I'll disable those tests via is-ci if needed. Any objections?

@molant can this be a problem in a scanner using chrome connector?

sarvaje commented on packages/connector-chrome/tests/events.ts:
```diff
@@ -295,7 +295,7 @@ test(`[${name}] Events`, async (t) => {
 
     // List of events that only have to be called once per execution
-    const singles = ['fetch::error', 'scan::start', 'scan::end', 'fetch::end::html'];
+    const singles = ['scan::start', 'scan::end', 'fetch::end::html'];
```
question: Why did you remove the `fetch::error`?

molant replied:
Because of https://github.com/webhintio/hint/issues/1398#issuecomment-429929512
`fetch::error` is no longer a single event, we receive it twice probably because of speculative downloading.

sarvaje commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -16,30 +16,30 @@ import { promisify } from 'util';
+import { getType } from 'hint/dist/src/lib/utils/content-type';
 import { debug as d } from 'hint/dist/src/lib/utils/debug';
-import * as logger from 'hint/dist/src/lib/utils/logging';
+// import * as logger from 'hint/dist/src/lib/utils/logging';
```
remove comment.

sarvaje commented on packages/utils-debugging-protocol-common/src/RequestResponse.ts:
```diff
@@ -0,0 +1,271 @@
+        this.logInfo(RequestStatus.loadingFailed);
+    }
+
+    public constructor(network: Crdp.NetworkClient, event: Crdp.Network.RequestWillBeSentEvent, overrideInvalidCert: boolean) {
```
IMO, `overrideInvalidCert` should be part of an `options` object. Just in case we need to pass something else in the future.

molant replied:
`RequestResponse` is meant to be used only by `utils-debugging-protocol`. If we need to add more things we can then switch to an `options` object without breaking.

antross commented on packages/utils-debugging-protocol-common/src/debugging-protocol-connector.ts:
```diff
@@ -247,7 +246,7 @@ export class Connector implements IConnector {
         debug(`Error found loading ${resource}:\n%O`, params);
 
-        const element: AsyncHTMLElement = (await this.getElementFromRequest(params.requestId))!;
+        const element: AsyncHTMLElement = (await this.getElementFromRequest(params.requestId, this._dom))!;
```
Ideally this should be `AsyncHTMLElement | null` (or simply omitted), but that requires updating the definition of `FetchError` to have `element: IAsyncHTMLElement | null`, which would be good to align it with the definition of `FetchEnd` that already allows `null`, but may be more than you want to do for this PR (as adding the `null` to `FetchError` could have ripple effects requiring changes in other packages).

molant replied:
Yeah, I tried that before submitting the changes but it seemed to be more changes than needed. I'm working on fixing the perf budget tests but I can look into this later.

molant replied:
I've fixed this.

molant commented:
All feedback should be in and hopefully test pass now. At the end I had to force the download of the favicon when running as `--headless` so there are a few more files changed to expose the flags used to open chrome.
If you can take another look it will be great.

sarvaje commented on packages/connector-chrome/tests/collect.ts:
```diff
@@ -55,11 +55,19 @@ const runTest = async (t: GenericTestContext<Context<any>>, ConnectorConstructor
     await connector.close();
 };
 
+test(`[${name}] The HTML is downloaded is present in a 'link' element with 'rel' attribute set to 'icon' `, async (t) => {
```
>The HTML ~is~ downloaded is present`?

molant commented:
@antross, I've modified a few new files since you last approved it. Can you please take a look?
@alrra do you want to take a look? I'd like to have this merged and published tomorrow Friday and before merging any of the breaking changes so we can do another big run test.
I know you are traveling so I can take care of that if you can't.

antross commented:
>  I've modified a few new files since you last approved it. Can you please take a look?

@molant Still looks good to me.

alrra commented on packages/utils-debugging-protocol-common/src/RequestResponse.ts:
```diff
@@ -0,0 +1,290 @@
+export class RequestResponse {
+    private _overrideInvalidCert: boolean;
+
+    /** The debugging protocol Network Client used to download the response body */
```
For consistency add `.` at the end, and and everywhere else.

alrra commented on packages/utils-debugging-protocol-common/src/RequestResponse.ts:
```diff
@@ -0,0 +1,290 @@
+
+    /** The final URL after redirects. Could be the same as `originalUrl`. */
+    public get finalUrl(): string {
+        // `willBeSent` gets updated with each redirect so this is the final requested url
```
> URL

alrra commented on packages/utils-debugging-protocol-common/src/RequestResponse.ts:
```diff
@@ -0,0 +1,290 @@
+                }
+            };
+
+
```
Remove extra empty line.
